classes/page_group.php

```diff
@@ -0,0 +1,114 @@
+<?php
+// This file is part of Moodle - https://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace tool_excimer;
+
+use core\persistent;
+
+/**
+ * Metadata for profile groups.
+ *
+ * @package   tool_excimer
+ * @author    Jason den Dulk <jasondendulk@catalyst-au.net>
+ * @copyright 2022, Catalyst IT
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class page_group extends persistent {
+
+    /** The name of the database table. */
+    public const TABLE = 'tool_excimer_page_groups';
+
+    /**
+     * Returns the fuzzy counts for duration.
+     *
+     * @return array
+     */
+    protected function get_fuzzydurationcounts(): array {
+        $raw = $this->raw_get('fuzzydurationcounts');
+        if (empty($raw)) {
+            return [];
+        }
+
+        return json_decode($raw, true);
+    }
+
+    /**
+     * Sets the fuzzy duration counts.
+     *
+     * @param array $data
+     */
+    protected function set_fuzzydurationcounts(array $data) {
+        $this->raw_set('fuzzydurationcounts', json_encode($data));
+    }
+
+    /**
+     * Records fuzzy count metadata about a page group.
+     *
+     * @param profile $profile The profile to pull the information from.
+     */
+    public static function record_fuzzy_counts(profile $profile) {
+
+        // Get the profile group record, creating a new one if one does not yet exist.
+        $month = userdate(time() - 360, '%Y%m'); // YYYYMM format.
+        $profilegroup = self::get_record(['name' => $profile->get('groupby'), 'month' => $month]);
+        if ($profilegroup === false) {
+            $profilegroup = new page_group(0, (object) ['name' => $profile->get('groupby'), 'month' => $month]);
+        }
+
+        $existing = $profilegroup->to_record();
+
+        // Fuzzy increment the count.
+        $current = $profilegroup->get('fuzzycount');
+        $new = manager::approximate_increment($current);
+        $profilegroup->set('fuzzycount', $new);
+
+        // Fuzzy increment the duration count.
+        $duration = $profile->get('duration');
+        $fuzzydurationcounts = $profilegroup->get('fuzzydurationcounts');
+        $fuzzydurationsum = $profilegroup->get('fuzzydurationsum');
+        $high = 1;
```

Comment:
this needs a comment to help grok

I have a feeling something isn't right with the fuzzydurationsum but I'll test it a bit more later and see.

Summing rather than counting will need to tweak the algorithm slightly, if you count 1 hit then you do the random logic once, but if you are summing say a value of 'this page took 10 seconds' then it would be equivalent to running a count of +1 but in a loop doing it 10 times so you have 10 times the chance of bumping the count up a factor of 2.

There is probably a more efficient way of probabilistic summing which doesn't involve a dumb loop. Some out loud thinking:

If the current sum is 2**4 = 8 and you want to add 8 then we know that 100% of the time it should bump up to to 2**5 = 16

If the current sum is 2**4 = 8 and you want to add 4 then we know that 50% of the time it should bump up to to 2**5 = 16

If the current sum is 2**4 = 8 and you want to add 2 then we know that 25% of the time it should bump up to to 2**5 = 16

If the current sum is 2**4 = 8 and you want to add 1024 then we know that 100% of the time we should practically ignore the current value and bump it directly to 2**10 = 1024

If the current sum is 2**9 = 512 and you want to add 1024 then we know that 50% of the time we should bump 2**10 = 1024

Adding n to m, should be the same as adding m to n (which sounds dumb saying out loud but worth being explicit here)

So I think we need a second algorithm helper function with takes 2 numbers like 8 and 1024 (which may not be factors of 2) and returns at minimum the maximum of the log of both numbers, and then some variable chance of an extra factor bump depending on how close they are.

I don't really understand the current code so maybe it is equivalent to this but not sure

Reply:
I presume you mean 2^4 = 8 and 2^10 = 1024?

Reply:
ah github mangled my comments, yes re-read that

Reply:
> I don't really understand the current code so maybe it is equivalent to this but not sure

It does nothing like it, but thanks for clarifying things.

Reply:
I found this a little too vague on what to do when the difference between the logs is 2 or greater, so I just put in a dumb loop.